### ISSUE TITLE
Unexpected Disconnection Due to Timer

### ISSUE THREAD

**Issue by VeMax** [weliem/blessed-android-coroutines#43]
Hi,

I’ve been experiencing an issue where the device disconnects shortly after establishing a connection. It seems that the disconnection is triggered by a connection timer, although this shouldn’t be happening. I use the latest 0.4.1 version.

• The connection process begins and successfully reaches the CONNECTED state (I use `centralManager.connectPeripheral()`).
• Shortly after, the startConnectionTimer triggers a disconnection due to a timeout, although the connection should be stable.
• The timer is set to disconnect the peripheral even if the connection was successful.

Here are the logs:
[disconnection-logs.txt](https://github.com/user-attachments/files/16391011/disconnection-logs.txt)

**Comment by VeMax** [comment]
I found the problem.

Steps to Reproduce:
	1. Ensure nRF Connect is connected to the target device.
	2. Use the Blessed library to connect to the same device using `central.connectPeripheral(central.getPeripheral("macAddress"))`
	3. Observe the logs for connection and disconnection events.
	
I added logs to find the problem
```kotlin
private fun startConnectionTimer(peripheral: BluetoothPeripheral) {
        cancelConnectionTimer()

        timeoutJob = scope.launch {
            Timber.i("TEST_TEST timeoutJob started") // logs
            delay(CONNECTION_TIMEOUT_IN_MS)
            Logger.e(TAG, "connection timeout, disconnecting '%s'", peripheral.name)
            disconnect()
            scope.launch {
                delay(50)
                bluetoothGatt?.let {
                    bluetoothGattCallback.onConnectionStateChange(bluetoothGatt, HciStatus.CONNECTION_FAILED_ESTABLISHMENT.value, BluetoothProfile.STATE_DISCONNECTED)
                }
            }
        }
    }
``` 
```kotlin
/**
     * This abstract class is used to implement BluetoothGatt callbacks.
     */
    private val bluetoothGattCallback: BluetoothGattCallback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            Timber.i("TEST_TEST cancelled: ${newState != BluetoothProfile.STATE_CONNECTING} " + // logs
                    "gatt=$gatt status=$status newstate=$newState") // logs
            if (newState != BluetoothProfile.STATE_CONNECTING) cancelConnectionTimer()
            val previousState = state
            state = newState

            val hciStatus = HciStatus.fromValue(status)
            if (hciStatus == HciStatus.SUCCESS) {
                when (newState) {
                    BluetoothProfile.STATE_CONNECTED -> successfullyConnected()
                    BluetoothProfile.STATE_DISCONNECTED -> successfullyDisconnected(previousState)
                    BluetoothProfile.STATE_DISCONNECTING -> {
                        Logger.d(TAG, "peripheral is disconnecting")
                        listener.disconnecting(this@BluetoothPeripheral)
                    }
                    BluetoothProfile.STATE_CONNECTING -> {
                        Logger.d(TAG, "peripheral is connecting")
                        listener.connecting(this@BluetoothPeripheral)
                    }
                    else -> Logger.e(TAG, "unknown state received")
                }
            } else {
                connectionStateChangeUnsuccessful(hciStatus, previousState, newState)
            }
        }
     }
``` 

The problem is that you cancel job before it's even started. And here you can see the logs
[disconnect-with-logs.txt](https://github.com/user-attachments/files/16391996/disconnect-with-logs.txt)

By the way, the problem appeared in version 0.3.2
<img width="1815" alt="image" src="https://github.com/user-attachments/assets/b4a3267f-2598-4076-8e41-47de0819c802">

Seems like this can fix the problem

```kotlin
 /**
     * Connect directly with the bluetooth device. This call will timeout in max 30 seconds (5 seconds on Samsung phones)
     */
    fun connect() {
        if (state == BluetoothProfile.STATE_DISCONNECTED) {
            scope.launch {
                delay(DIRECT_CONNECTION_DELAY_IN_MS)
                Logger.d(TAG, "connect to '%s' (%s) using TRANSPORT_LE", name, address)
                registerBondingBroadcastReceivers()
                discoveryStarted = false
                bluetoothGatt = try {
                    device.connectGatt(context, false, bluetoothGattCallback, BluetoothDevice.TRANSPORT_LE)
                } catch (e: SecurityException) {
                    Logger.d(TAG, "exception")
                    null
                }
                bluetoothGatt?.let {
                    connectTimestamp = SystemClock.elapsedRealtime()
                    startConnectionTimer(this@BluetoothPeripheral) // first start timer, then change connection state
                    bluetoothGattCallback.onConnectionStateChange(it, HciStatus.SUCCESS.value, BluetoothProfile.STATE_CONNECTING)
                }
            }
        } else {
            Logger.e(TAG, "peripheral '%s' not yet disconnected, will not connect", name)
        }
    }
```

**Comment by weliem** [comment]
Indeed, there is a possible race conditions. Fixed by 67cedddec56269e473adb595ad00b07cc9220ceb
